### DISCUSSION TITLE
Trouble when upgrading from prisma 3.9.2 to 4.10.1

### DISCUSSION THREAD

**Discussion by Tseredian** [prisma/prisma#18112] [Q&A] [answered]
Hello,

I'm having trouble when upgrading the prisma version in my project. there are no schema errors and upgrading is smooth. But on some update queries where we had interactive transactions, it throws error: "All elements of the array need to be Prisma Client promises. Hint: Please make sure you are not awaiting the Prisma client calls you intended to pass in the $transaction function." 

I have following code inside the function:

```
let transactions = [];
transactions.push(
                this.prismaService.$transaction(async (tx) => {
                    let update = await tx.table1.update({
                        where: {
                            id: id
                        },
                        data: {
                            //some data
                        }
                    })
                    if(some condition) {
                        await tx.table2.update();
                    }
                }
            )
//some other transactions are added
return await this.prismaService.$transaction(transactions);
```

So I guess what I'm asking now, is since this way doesn't work anymore,how can I combine interactive transactions,with regular queries to execute both of them transactionally?

Any help or insights are greatly appreciated, I've been stuck on this for quite some time now... 

Apologies if question is not formatted well or if my explanation is not well worded, it's my first time asking question on this platform

Thank you

**Comment by millsp** [comment, marked as answer]
Hey @Tseredian :wave: The error is correct, you cannot nest a `$transaction` in a batch `$transaction` this way. So if that worked, it was most likely a bug because this error message exists for a long time. That said, I am here to help you!

First of all, a `PrismaPromise` represents anything that will issue a query. So that is all the raw queries as well as all the model queries. `PrismaPromises` can be chained on (ie. fluent API) on models, but most importantly they provide a way to inject them with a transaction context - be it in interactive transactions or in batch transactions. `$transaction` does not return a `PrismaPromise`, so you cannot nest them.

In other words, because your queries are already being run in a transaction, you don't need to pass them to the `$transaction` call again. So if I understood your code correctly, you just want to execute them one after another? In that case, this should work for you:

```ts
let transactions = [];
transactions.push(
                this.prismaService.$transaction(async (tx) => {
                    let update = await tx.table1.update({
                        where: {
                            id: id
                        },
                        data: {
                            //some data
                        }
                    })
                    if(some condition) {
                        await tx.table2.update();
                    }
                }
            )
//some other transactions are added

let results = [];
for (const tx of transactions) {
    results.push(await tx);
}

return results;
```

**Reply by Tseredian** [reply]
thank you for the answer.

So batch transactions and interactive transactions are not possible to execute together transactionally ? And If I want whole operation to be atomic I have to choose one or the other?

**Reply by millsp** [reply]
> So batch transactions and interactive transactions are not possible to execute together transactionally

Indeed it is not possible to do it this way, but you are welcome to open a feature request.

> And If I want whole operation to be atomic I have to choose one or the other?

So you want all the operations to be atomic, kinda merging them into a single transaction? In that case, what you could do is to push all your transaction callbacks into an array, and then execute all these stored callbacks in a single interactive transaction.

**Reply by Tseredian** [reply]
Yes that's the ideal scenario. I found a convenient way to chain interactive transactions in one of the discussions here which saved me from lot of refactoring

https://github.com/prisma/prisma/discussions/12373#discussioncomment-3909212

 so I guess at the moment I'm good.

Thank you for your time, at least now I know it's not possible and I can stop searching

**Reply by millsp** [reply]
You're very welcome. Happy to help you again in the future.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
